### ISSUE TITLE
Run tests using real sentry dsn

### ISSUE THREAD

**Issue by damiencaselli** [damiencaselli/ember-cli-sentry#21]
What would really be awesome is to be able to run the tests against a working sentry config, which is unfortunately not possible since this part is hosted by Sentry.

I wish I personally had the money to provide a working _dsn_, but I don't :smile: 

May be @mattrobenolt could give us some hints about running the tests more efficiently (testing _dsn_ may be).

**Comment by mattrobenolt** [comment]
This has come up in the past, but we don't really have a good answer.

The best we came up with was a potential JSON Schema to use for testing as discussed in https://github.com/getsentry/sentry/issues/1018, but we never made progress on that.

fwiw, none of our official clients today test against a running Sentry instance for automated tests. We will manually test against a running instance occasionally to validate some behavior, but never automated.

If you just want a test account, you can sign up for a free plan, and if you need more than what the free plan provides, we can hook you up.

**Comment by damiencaselli** [comment]
@mattrobenolt I see.

May be having a dedicated testing key owned by the sentry team could fix that, since you can handle throttling and quotas.

Although I do have an account on Getsentry and I'd gladly share a key for testing purposes, 250 events per day can be reached rather quickly if a couple of people are using it.

Mocking the API could also be nice, but requires more work on every side (sentry team, contributors…).

My call is to let people run tests against their own keys, at least until a better solution is found.

> I provide the car, you bring the oil.

**Comment by asbjornu** [comment]
@damiencaselli I also believe that until Sentry has a JSON Schema that most use cases can be automatically tested against, the Sentry team (@mattrobenolt, @dcramer) should provide a DSN that can be used across all Raven clients to perform automatic testing.

I've envisioned a devious plan in getsentry/raven-csharp#61 to self-host the Python API in the C# test suite (fire up a Python background process listening on a specific port), but if we could just do integration tests directly against the public Sentry API, that would be much, much simpler for all Raven client developers.

**Comment by mattrobenolt** [comment]
The biggest problem with running automated tests against our hosted infrastructure is going to be that you're going to get inconsistent results.

For one, your requests would be rated limited almost immediately, so you'll start getting 429 responses and you'll likely not know that what you're sending is working anyways. Regardless of whatever plan you're on, we ultimately rate limit at IP level and whatnot to protect ourselves, which does not favor well for automated testing.

I'm not opposed to handing out a key for integration testing, I just believe those tests should be done manually, and not a part of an automated test suite because as I said, it's unlikely that's going to yield the results you're looking for.

Just so I understand so we can design a solution, what exactly are you looking for in an integration test with our API? The response that's kicked back isn't very useful in determining if it worked or not since we do most validation asynchronously. Basically, as long as you've sent an ok looking packet with a valid DSN, we'll kick back a 200 OK with a little bit of JSON (or a transparent gif for JavaScript). Other than that, we wouldn't be surfacing much other feedback from that API. In my opinion, you'd also want control over the responses kicked back so you can validate certain behaviors manually. For example, how does the client handle a 429 response vs 200 vs a network failure vs whatever else?

With that said, as mentioned, something like a JSON schema to validate against is more probable and more likely to help with the situation since that can be a more extensive test to make sure everything is correctly formatted. There are probably some other solutions that we haven't explored, but I'm not sure what they could be off the top of my head.

I'd take a look at the test suites we use in raven-python, raven-js, raven-ruby, and raven-go as examples for how we do it ourselves internally. None of these touch our public Sentry servers, or any Sentry server.

tl;dr, if you still want a key, let me know, but it won't be without restrictions that we can't remove. Running an automated test suite against our API is going to hit rate limiting issues if you're trying to do any reasonable number of tests against it. Not sure how people develop things that work with other public API's, but I know they don't typically hit the services they're writing tests for. We should look for inspiration from Twitter client libraries, GitHub libraries, etc, and see how they handle it. My guess is they use mocks, but I don't have much personal experience.

We're willing to help, I just want to make sure we're accomplishing the shared goals here. :)

**Comment by damiencaselli** [comment]
@mattrobenolt thanks for the insights.

With ember-cli-sentry, Sentry integration only consists in having a valid _dsn_ to ensure config and install are properly done.
Logging failing does not concern the add-on, its role is just to make sure that Raven gets properly set up given a config, valid or not.

Nothing else needs to be done since in-depth tests are performed by Raven.js.

I believe a valid key with 0 quotas would in fact do the trick, it just needs to be setup.

This was the lazy part :smile: 

This is a totally different matter for @asbjornu and everyone implementing a library really using Sentry API, since they need to replicate success and errors.
Since Sentry code is available, may be running tests against a private instance would be the most reliable way of doing it, provided one can build the exact same .

It makes tests a bit more difficult to run since one has to run their own code, but I guess a Docker container or any orchestrator recipe (saltstack…) might be enough.

**Comment by damiencaselli** [comment]
@mattrobenolt Ahhh I just noticed registering on GetSentry with an email aliases was fine.

I think I'll create a new account for that purpose, it would perfectly suit my needs.

**Comment by asbjornu** [comment]
@mattrobenolt I see. Thanks for all the insight. An overview on how the different Raven clients handle this for other platforms would be very valuable.

**Comment by damiencaselli** [comment]
I've added a key to the test application, and now some Raven logging should be displayed to the console when one tests the add-on.

That being said, if everyone test the add-on, quotas will be exceeded by far since the account is a limited one. Not that I bother since I use it only for this add-on purpose, but one may get 4xx from the services.

You have been warnd :smile:
